function_calling/src/function_calling.py

```diff
@@ -0,0 +1,198 @@
+import os
+import re
+import sys
+import json
+from pprint import pprint
+from dotenv import load_dotenv
+from typing import Optional, Union
+from langchain_community.llms.sambanova import SambaStudio, Sambaverse
+from langchain_core.output_parsers import JsonOutputParser
+from langchain_core.prompts import PromptTemplate, ChatPromptTemplate
+from langchain_core.pydantic_v1 import BaseModel, Field
+from langchain_core.tools import Tool
+from langchain_core.runnables import RunnableLambda
+from langchain_core.messages.human import HumanMessage
+from langchain_core.messages.ai import AIMessage
+from langchain_core.messages.tool import ToolMessage
+
+current_dir = os.path.dirname(os.path.abspath(__file__))
+kit_dir = os.path.abspath(os.path.join(current_dir, ".."))
+repo_dir = os.path.abspath(os.path.join(kit_dir, ".."))
+sys.path.append(kit_dir)
+sys.path.append(repo_dir)
+
+
+load_dotenv(os.path.join(repo_dir, ".env")) 
+
+
+FUNCTION_CALLING_SYSTEM_PROMPT = """you are an helpful assistant and you have access to the following tools:
+
+{tools}
+
+You must always select one or more of the above tools and answer with only a list of JSON objects matching the following schema:
+
+```json
+[{{
+  "tool": <name of the selected tool>,
+  "tool_input": <parameters for the selected tool, matching the tool's JSON schema>
+}}]
+```
+
+Think step by step
+Do not call a tool if the input depends on another tool output you dont have yet
+Do not try to answer until you get tools output, if you dont have an answer yet you can continue calling tools
```

Comment:
Do not try to answer until you get at least one tools output (or all the tools output?).
if you do not have an answer yet, you can continue calling tools until you do.

Reply:
fixed